website_event_sale_seat_reserve/models/event_registration.py

```diff
@@ -0,0 +1,21 @@
+# Copyright 2024 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl-3.0).
+
+from odoo import api, models
+
+
+class EventRegistration(models.Model):
+    _inherit = "event.registration"
+
+    def _need_pre_reservation(self):
+        need_pre_reservation = self.sale_order_id and self.sale_order_line_id
+
+        return need_pre_reservation and self.visitor_id
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        registrations = super(EventRegistration, self).create(vals_list)
+        for registration in registrations:
+            if registration._need_pre_reservation():
+                registration.action_set_reserved()
+        return registrations
```

Comment:
For readability, I suggest something like this:

```py
    def _auto_reserve_registrations(self):
        to_reserve = self.filtered(
            lambda rec: (
                rec.state == 'draft'
                and rec.sale_order_line_id
                and rec.visitor_id
            )
        )
        to_reserve.action_set_reserved()

    @api.model_create_multi
    def create(self, vals_list):
        registrations = super().create(vals_list)
        registrations._auto_reserve_registrations()
        return registrations
```

Notice I added also a check for `state == 'draft'`.

As I recall, in core, when visitors register themselves on free events they get automatically confirmed registrations. So, just in case, it wouldn't hurt to check the state in case they are already confirmed